File: articles/cheminformatics-molecular-data-science/julia/cheminformatics_kernel.jl
# Cheminformatics numerical kernel in Julia.

function tanimoto(a, b, c)
    return c / (a + b - c)
end

function pic50(ic50_nm)
    ic50_m = ic50_nm * 1.0e-9
    return -log10(ic50_m)
end

function euclidean_distance(x, y)
    return sqrt(sum((x .- y).^2))
end

println("tanimoto=", round(tanimoto(5.0, 4.0, 3.0), digits=6))
println("pIC50=", round(pic50(50.0), digits=6))
println("distance=", round(euclidean_distance([1.0, 2.0, 3.0], [1.5, 2.5, 4.0]), digits=6))

File: articles/cheminformatics-molecular-data-science/fortran/cheminformatics_kernel.f90
program cheminformatics_kernel
  implicit none

  real(8) :: tanimoto, pic50, distance
  real(8) :: ic50_nm, ic50_m

  tanimoto = 3.0d0 / (5.0d0 + 4.0d0 - 3.0d0)

  ic50_nm = 50.0d0
  ic50_m = ic50_nm * 1.0d-9
  pic50 = -log10(ic50_m)

  distance = sqrt((1.0d0 - 1.5d0)**2 + (2.0d0 - 2.5d0)**2 + (3.0d0 - 4.0d0)**2)

  print *, "tanimoto=", tanimoto
  print *, "pIC50=", pic50
  print *, "distance=", distance

end program cheminformatics_kernel
